vms/platformvm/warp/benchmarks/benchmarks_test.go

```diff
@@ -0,0 +1,381 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package warpbench
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/choices"
+	"github.com/ava-labs/avalanchego/utils/crypto/bls"
+	"github.com/ava-labs/avalanchego/utils/crypto/secp256k1"
+	"github.com/ava-labs/avalanchego/utils/timer/mockable"
+	"github.com/ava-labs/avalanchego/vms/components/avax"
+	"github.com/ava-labs/avalanchego/vms/platformvm/blocks"
+	"github.com/ava-labs/avalanchego/vms/platformvm/reward"
+	"github.com/ava-labs/avalanchego/vms/platformvm/signer"
+	"github.com/ava-labs/avalanchego/vms/platformvm/state"
+	"github.com/ava-labs/avalanchego/vms/platformvm/status"
+	"github.com/ava-labs/avalanchego/vms/platformvm/txs"
+	"github.com/ava-labs/avalanchego/vms/platformvm/warp"
+	"github.com/ava-labs/avalanchego/vms/secp256k1fx"
+	"github.com/stretchr/testify/require"
+)
+
+// To profile a specific benchmark, say BenchmarkGetCanonicalValidatorSetByDepth, run:
+// go test -run=^$ -bench ^BenchmarkGetCanonicalValidatorSetByDepth$ github.com/ava-labs/avalanchego/vms/platformvm/warp/benchmarks -cpuprofile=cpuprofile.out
+//
+// To analyze results run:
+// go tool pprof -web cpuprofile.out
+
+func BenchmarkGetCanonicalValidatorSetBySize(b *testing.B) {
+	b.StopTimer()
+	require := require.New(b)
+
+	// Prepare the bench environment (don't time it)
+	subnetID := ids.GenerateTestID()
+	env := newEnvironment(b, subnetID)
+	defer func() {
+		require.NoError(shutdownEnvironment(env))
+	}()
+
+	pChainHeight := uint64(1)
+	numNodes := 10_000
+
+	// Store [numNodes] validators in state. They are all at the same height [pChainHeight]
+	diff, err := state.NewDiff(env.state.GetLastAccepted(), env.blkManager)
+	require.NoError(err)
+	for i := 0; i < numNodes; i++ {
+		nodeID := ids.GenerateTestNodeID()
+
+		// create primary network validator tx
+		addPrimaryValidatorTx, err := env.txBuilder.NewAddValidatorTx(
+			env.config.MinValidatorStake,
+			uint64(env.state.GetTimestamp().Unix()),
+			uint64(mockable.MaxTime.Unix()),
+			nodeID,
+			ids.ShortEmpty, // reward address
+			reward.PercentDenominator,
+			[]*secp256k1.PrivateKey{preFundedKeys[0]},
+			ids.ShortEmpty,
+		)
+		require.NoError(err)
+
+		// store corresponding primaryStaker in the diff
+		primaryStaker, err := state.NewCurrentStaker(
+			addPrimaryValidatorTx.ID(),
+			addPrimaryValidatorTx.Unsigned.(*txs.AddValidatorTx),
+			10000, // potential reward
+		)
+		require.NoError(err)
+		diff.PutCurrentValidator(primaryStaker)
+		diff.AddTx(addPrimaryValidatorTx, status.Committed)
+
+		// create subnet validator tx
+		blsPrivateKey, err := bls.NewSecretKey()
+		require.NoError(err)
+
+		uPermissionlessValidatorTx := &txs.AddPermissionlessValidatorTx{
```

Comment:
could we move this large initialization to a separate fn and share it between both benchmarks?